### PR TITLE
[ORTModule] ATen Efficient Attention and Triton Flash Attention

### PR DESCRIPTION
**Pull request by er3x3**
This PR is to support efficient attention and flash attention in ORTModule, including:
- Use ATen to call efficient attention, which requires PyTorch 2.2.0 dev or newer. ORTMODULE_USE_EFFICIENT_ATTENTION=1 to enable.
- Integrate Triton Flash attention, which requires triton==2.0.0.dev20221202. Need A100 or H100. ORTMODULE_USE_FLASH_ATTENTION=1 to enable.
- A python transformer tool to match sub-graph by config and write transformer quickly.

Current transformers supports attention mask for both efficient attn and flash attn, and dropout for efficient attn only. To support more training scenarios (such as causal mask in GPT2), more transformers need to be added.

The feature is guarded by system environment variables, it won't effect any current behavior if not enabled. Since it requires specific PyTorch/Triton versions, related tests is not added for now.

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/transformers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+
+import os
+
+_all_transformers = []
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/transformers/__init__.py:8]: can we rename to graph_optimizer or something to make it clear what this does... transformer is overloaded term.

File: onnxruntime/contrib_ops/cpu/aten_ops/aten_op.cc [hunk 2]
```diff
@@ -32,8 +32,10 @@ Status ATen::Compute(OpKernelContext* p_ctx) const {
   aten_ops::ATenOperatorExecutor::Instance()(op_name_, overload_name_, input_size, dlpack_inputs.get(), output_size,
                                              dlpack_outputs.get());
   for (size_t i = 0; i < output_size; ++i) {
-    ORT_RETURN_IF_ERROR(
-        p_ctx_internal->SetOutputMLValue(static_cast<int>(i), dlpack::DlpackToOrtValue(dlpack_outputs[i])));
+    if (dlpack_outputs[i]) {
```
Comment (askhade) [inline onnxruntime/contrib_ops/cpu/aten_ops/aten_op.cc:35]: is the else case valid? Do we expect some outputs coming from dlpack and some not?

**Review by askhade — APPROVED** [review]
